### PR TITLE
add feature to allow adding custom liveness paths to dropwizard health

### PR DESCRIPTION
**Pull request by richeyh**
###### Problem:
This is written up in https://github.com/dropwizard/dropwizard/issues/5518

TLDR:
Need a way to break up alive and readiness checks on separate URL's

###### Solution:
This solution modifies the `ServletHealthResponderFactory.java` to have a new optional configurable field `livenessCheckUrlPaths`. This field is a list of string URLS that should only return the results of the alive checks. This makes it a non breaking change as the probably better solution would be to add the field to the `DefaultHealthFactory` and rename the `healthCheckUrlPaths` to be more clear. However, this change would be breaking and probably best served for the 3.x or 4.x release. 

###### Result:
There is now a new option to give consumers further control of their health checks.

### CODE REVIEW

**Comment by pstackle** [comment]
@richeyh Can you explain in more detail your use case such that the current behavior doesn't support it? 

The current approach does allow you to configure separate liveness and readiness probes for a Kubernetes pod via the query string parameter you mentioned:
```
# use for readiness probe
https://<hostname>:<port>/health-check?type=ready

# use for liveness probe
https://<hostname>:<port>/health-check?type=alive
```

This approach has the benefit of making it easier to support additional types of health probes that come up in the future as well. 

The purpose of the `healthCheckUrlPaths` was not to have different paths return different results or be tied to a specific type of health check, but rather to support an app that needed to migrate the health check URL used from one location to another and allowing to configure multiple enables that to happen seamlessly. In most cases, there is no need to ever configure anything beyond the default value and even in the case of a migration, one wouldn't typically need more than two.

The reason that the health check call returns the readiness result by default is that it was initially the only one that was supported, the `alive` support came later, and it is also the most commonly used one.

**Comment by richeyh** [comment]
Yep, so right now with the default being liveness it essentially means unless you use query params you can't get two distinct urls. Our use case is we have our health check library that currently supports this and we want to move to the health check library that's part of core. However, for us, it's important to keep the urls the same which would mean the ability to configure a new url for liveness alone(not just health as it currently stands). As this seemed like a good use case for teams vs using a query param.  Also when you look at other frameworks like spring boot, which allows url config per type, it seemed like this would be a welcome addition.

**Comment by pstackle** [comment]
Can you explain this in more detail?
> As this seemed like a good use case for teams vs using a query param.

I think I'm still missing something with respect to what the problem is with query parameters.

**Comment by richeyh** [comment]
> Can you explain this in more detail?

Yep so right now you can only configure `healthCheckUrlPaths`  this corresponds to endpoints that only function for readiness of the application. There is no config currently to add a distinct endpoint for only livenesss. The use of the query param is a workaround but it still does not give applications a way to have two endpoints that serve different purposes, by default every endpoint I configure would have to be for readiness. Now if we go look at other frameworks we see in [spring boot](https://www.baeldung.com/spring-boot-actuators#6-health-groups) they are distinct configurable endpoints to further clarify the difference in readiness. 

 My full use case to match the current behavior and runtime constraints of my environment is I need my app to serve readiness on `/management/health/readiness` and liveness on `/management/health/liveness`. Yes I can probably achieve this with a redirect filter, however it seems worthwhile that other users would want the pattern of configurability for the endpoints that are determining the liveness and readiness of their application.

**Comment by pstackle** [comment]
> My full use case to match the current behavior and runtime constraints of my environment is I need my app to serve readiness on /management/health/readiness and liveness on /management/health/liveness.

If your app is deployed in Kubernetes, what prevents you from configuring the liveness and readiness probe urls as part of your pod spec?

```
    livenessProbe:
      httpGet:
        path: /health-check?type=alive
    readinessProbe:
      httpGet:
        path: /health-check?type=ready
```

As I mentioned earlier, trying to support a list of URLs per status type doesn't seem like that would scale very well to me. 

Is the hangup related to the use of query params or the fact that you don't have control of the status type value? 

Hypothetically, if the following worked, would that somehow work better for you or would you still be blocked?
```
alive: /health-check/alive
ready: /health-check/ready

general pattern: /[configured url]/[status type]
```

**Comment by richeyh** [comment]
> If your app is deployed in Kubernetes, what prevents you from configuring the liveness and readiness probe urls as part of your pod spec?

We have internal contracts for all our applications to match. Right now our node apps and our spring boot apps can conform to the contract. Our Dropwizard apps with our bundle can also conform. However, without a change in the core health check bundle, it will not be able to conform. 

> As I mentioned earlier, trying to support a list of URLs per status type doesn't seem like that would scale very well to me.

Do you mean there would be more than just the two status types? Are you looking to adopt something like health check groups? My understanding is that there would only be two groups/types of health checks alive and ready. 

> Hypothetically, if the following worked, would that somehow work better for you or would you still be blocked?

It's closer, we would also want to be able to customize the name of the groups or status type from your example.

**Comment by pstackle** [comment]
To close the loop on this:
It sounds like your requirements are pretty specific to the exact strings you need for your probe urls, so doing your own redirect filter is probably the better approach. 

Trying to support a list of URLs per status type wouldn't scale very well as Kubernetes expands the types of probes that it supports.

Given the above, I'm going to close this PR. If you strongly disagree, please open a Discussion item to allow for further discussion about the use case.
